<a href="https://colab.research.google.com/github/LatiefDataVisionary/numpy-with-w3schools/blob/main/numpy_ufuncs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. ufuncs Intro**

Without ufunc, we can use Python's built-in zip() method:



In [39]:
x = [1,2,3,4]
y = [4,5,6,7]
z=[]

for i, j in zip(x,y):
  z.append(i+j)
z

[5, 7, 9, 11]

With ufunc, we can use the add() function:



In [40]:
import numpy as np

x = [1, 2, 3, 4]
y = [4, 5, 6, 7]
z = np.add(x,y)
z

array([ 5,  7,  9, 11])

# **2. ufuncs Create Function**

Create your own ufunc for addition:



In [41]:
def myadd(x,y):
  return x+y

myadd = np.frompyfunc(myadd, 2, 1)
myadd([1,2,3,4], [5,6,7,8])

array([6, 8, 10, 12], dtype=object)

Check if a function is a ufunc:



In [42]:
type(np.add)

numpy.ufunc

Check the type of another function: concatenate():



In [43]:
type(np.concatenate)

numpy._ArrayFunctionDispatcher

Use an if statement to check if the function is a ufunc or not:



In [44]:
if type(np.add) == np.ufunc:
  print('add is ufunc')
else:
  print('add is not ufunc')

add is ufunc


# **3. Simple Arithmetic**


Add the values in arr1 to the values in arr2:



In [45]:
arr1 = np.array([10, 11, 12, 13, 14, 15])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.add(arr1, arr2)
newarr

array([30, 32, 34, 36, 38, 40])

Subtract the values in arr2 from the values in arr1:



In [46]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.subtract(arr1, arr2)
newarr

array([-10,  -1,   8,  17,  26,  35])

Multiply the values in arr1 with the values in arr2:



In [47]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.multiply(arr1, arr2)
newarr

array([ 200,  420,  660,  920, 1200, 1500])

Divide the values in arr1 with the values in arr2:



In [48]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 10, 8, 2, 33])

newarr = np.round(np.divide(arr1, arr2), 2)
newarr

array([ 3.33,  4.  ,  3.  ,  5.  , 25.  ,  1.82])

Raise the valules in arr1 to the power of values in arr2:



In [49]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 6, 8, 2, 33])

newarr = np.power(arr1, arr2)
newarr

array([         1000,       3200000,     729000000, 6553600000000,
                2500,             0])

Both the mod() and the remainder() functions return the remainder of the values in the first array corresponding to the values in the second array, and return the results in a new array.



In [50]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.mod(arr1, arr2)
newarr

array([ 1,  6,  3,  0,  0, 27])

In [51]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.remainder(arr1, arr2)
newarr

array([ 1,  6,  3,  0,  0, 27])

In [52]:
# Contoh dengan bilangan negatif
arr1 = np.array([-10, 10])
arr2 = np.array([3, -3])

print("Modulus (mod):", np.mod(arr1, arr2))
print("Remainder (remainder):", np.remainder(arr1, arr2))

Modulus (mod): [ 2 -2]
Remainder (remainder): [ 2 -2]


Return the quotient and mod:



In [53]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.divmod(arr1, arr2)
newarr

(array([ 3,  2,  3,  5, 25,  1]), array([ 1,  6,  3,  0,  0, 27]))

Return the absolute values:



In [54]:
arr = np.array([-1, -2, 1, 2, 3, -4])
newarr = np.absolute(arr)
newarr

array([1, 2, 1, 2, 3, 4])

In [55]:
arr = np.array([-1, -2, 1, 2, 3, -4])
newarr = np.abs(arr)
newarr

array([1, 2, 1, 2, 3, 4])

# **4. ufunc Rounding Decimals**


Truncate elements of following array:



In [56]:
arr = np.trunc([-3.1666, 3.6667])
arr

array([-3.,  3.])

Same example, using fix():



In [57]:
arr = np.fix([-3.6667, 3.6667])
arr

array([-3.,  3.])

Round off 3.1666 to 2 decimal places:



In [58]:
arr = np.around(3.1666, 2)
arr

3.17

Floor the elements of following array:



In [59]:
arr = np.floor([-3.1666, 3.6667])
arr

array([-4.,  3.])

Ceil the elements of following array:



In [60]:
arr = np.ceil([-3.1666, 3.6667])
arr

array([-3.,  4.])

# **5. ufunc NumPy Logs**


Find log at base 2 of all elements of following array:



In [61]:
arr = np.arange(1, 10)
print(arr)
print(np.log2(arr))

[1 2 3 4 5 6 7 8 9]
[0.         1.         1.5849625  2.         2.32192809 2.5849625
 2.80735492 3.         3.169925  ]


Find log at base 10 of all elements of following array:



In [62]:
arr = np.arange(1, 11)
print(arr)
print(np.log10(arr))

[ 1  2  3  4  5  6  7  8  9 10]
[0.         0.30103    0.47712125 0.60205999 0.69897    0.77815125
 0.84509804 0.90308999 0.95424251 1.        ]


Find log at base e of all elements of following array:



In [63]:
arr = np.arange(1, 10)
np.log(arr)

array([0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
       1.79175947, 1.94591015, 2.07944154, 2.19722458])

Log at Any Base


In [64]:
from math import log

nplog = np.frompyfunc(log, 2, 1)
nplog(100, 15)

1.7005483074552052

Menghitung logaritma 100 dengan basis 15.
Hasilnya adalah sekitar 1.7005483074552052, karena
15
1.7005
≈
100
15
1.7005
 ≈100

# **6. NumPy Summations**


Sum the values in arr1 and the values in arr2:



In [65]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([1, 2, 3])

newarr = np.sum([arr1, arr2])
newarr

12

Perform summation in the following array over 1st axis:



In [66]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([1, 2, 3])

newarr_ax1 = np.sum([arr1, arr2], axis=1)
print(newarr_ax1)

[6 6]


Perform cummulative summation in the following array:



In [67]:
arr = np.array([1, 2, 3])
newarr = np.cumsum(arr)
newarr

array([1, 3, 6])

Cummulative sum means partially adding the elements in array.

E.g. The partial sum of [1, 2, 3, 4] would be [1, 1+2, 1+2+3, 1+2+3+4] = [1, 3, 6, 10].

Perfom partial sum with the cumsum() function.

# **7. NumPy Products**


Find the product of the elements of this array:



In [68]:
arr = np.array([1, 2, 3, 4])
x = np.prod(arr)
x

# Returns: 24 because 1*2*3*4 = 24

24

Find the product of the elements of two arrays:



In [69]:
arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

x = np.prod([arr1, arr2])
x

# Returns: 40320 because 1*2*3*4*5*6*7*8 = 40320

40320

Perform summation in the following array over 1st axis:



In [70]:
arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

newarr = np.prod([arr1, arr2], axis=1)
newarr

array([  24, 1680])

Take cummulative product of all elements for following array:



In [71]:
# E.g. The partial product of [1, 2, 3, 4] is [1, 1*2, 1*2*3, 1*2*3*4] = [1, 2, 6, 24]

arr = np.array([5, 6, 7, 8])
newarr = np.cumprod(arr)
newarr

array([   5,   30,  210, 1680])

# **8. NumPy Differences**


Compute discrete difference of the following array:



In [72]:
# E.g. for [1, 2, 3, 4], the discrete difference would be [2-1, 3-2, 4-3] = [1, 1, 1]

arr = np.array([10, 15, 25, 5])
newarr = np.diff(arr)
newarr

# Returns: [5 10 -20] because 15-10=5, 25-15=10, and 5-25=-20

array([  5,  10, -20])

We can perform this operation repeatedly by giving parameter `n`.

E.g. for [1, 2, 3, 4], the discrete difference with n = 2 would be [2-1, 3-2, 4-3] = [1, 1, 1] , then, since n=2, we will do it once more, with the new result: [1-1, 1-1] = [0, 0]

Compute discrete difference of the following array twice:



In [73]:
arr = np.array([10, 15, 25, 5])
newarr = np.diff(arr, n=2)
newarr

array([  5, -30])

**Returns**: [5 -30] because: 15-10=5, 25-15=10, and 5-25=-20 AND 10-5=5 and -20-10=-30



# **9. NumPy LCM (Lowest Common Multiple)**

Find the LCM of the following two numbers:



In [74]:
num1 = 4
num2 = 6

x = np.lcm(num1, num2)
x

12

**Returns**: 12 because that is the lowest common multiple of both numbers (4x3=12 and 6x2=12).



Find the LCM of the values of the following array:



In [75]:
arr = np.array([3, 6, 9])
x = np.lcm.reduce(arr)
x

18

The reduce() method will use the ufunc, in this case the lcm() function, on each element, and reduce the array by one dimension.



Find the LCM of all values of an array where the array contains all integers from 1 to 10:



In [76]:
arr = np.arange(1,11)
x = np.lcm.reduce(arr)
print(arr)
print(x)

[ 1  2  3  4  5  6  7  8  9 10]
2520


# **10. NumPy GCD (Greatest Common Denominator)**


Find the HCF of the following two numbers:



In [77]:
num1 = 6
num2 = 9
x = np.gcd(num1, num2)
x

3

**Returns**: 3 because that is the highest number both numbers can be divided by (6/3=2 and 9/3=3).



Find the GCD for all of the numbers in the following array:



In [79]:
arr = np.array([20, 8, 32, 36, 16])
x = np.gcd.reduce(arr)
x

4

**Returns**: 4 because that is the highest number all values can be divided by.



# **11. NumPy Trigonometric Functions**


Find sine value of PI/2:



In [84]:
x = np.pi/2
result = np.sin(x)
print(x)
print(result)

1.5707963267948966
1.0


Find sine values for all of the values in arr:



In [85]:
arr = np.array([np.pi/2, np.pi/3, np.pi/4, np.pi/5])
x = np.sin(arr)
x

array([1.        , 0.8660254 , 0.70710678, 0.58778525])

Convert all of the values in following array arr to radians:



In [86]:
arr = np.array([90, 180, 270, 360])
x = np.deg2rad(arr)
x

array([1.57079633, 3.14159265, 4.71238898, 6.28318531])

Convert all of the values in following array arr to degrees:



In [87]:
arr = np.array([np.pi/2, np.pi, 1.5*np.pi, 2*np.pi])
x = np.rad2deg(arr)
x

array([ 90., 180., 270., 360.])

Find the angle of 1.0:



In [88]:
x = np.arcsin(1.0)
x

1.5707963267948966

Find the angle for all of the sine values in the array



In [90]:
arr = np.array([1, -1, 0.1])
x = np.arcsin(arr)
x

array([ 1.57079633, -1.57079633,  0.10016742])

Find the hypotenues for 4 base and 3 perpendicular:



In [92]:
base, perpendicular = 3, 4
x = np.hypot(base, perpendicular)
int(x)

5

# **12. NumPy Hyperbolic Functions**


Find sinh value of PI/2:



In [93]:
x = np.sinh(np.pi/2)
x

2.3012989023072947

Find cosh values for all of the values in arr:



In [94]:
arr = np.array([np.pi/2, np.pi/3, np.pi/4, np.pi/5])
x = np.cosh(arr)
x

array([2.50917848, 1.60028686, 1.32460909, 1.20397209])

Find the angle of 1.0:



In [95]:
x = np.arcsinh(1.0)
x

0.881373587019543

Find the angle for all of the tanh values in array:



In [96]:
arr = np.array([0.1, 0.2, 0.5])
x = np.arctanh(arr)
x

array([0.10033535, 0.20273255, 0.54930614])

# **13. NumPy Set Operations**


Convert following array with repeated elements to a set:



In [97]:
arr = np.array([1, 1, 1, 2, 3, 4, 5, 5, 6, 7])
x = np.unique(arr)
x

array([1, 2, 3, 4, 5, 6, 7])

Find union of the following two set arrays:



In [98]:
arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([3, 4, 5, 6])

newarr = np.union1d(arr1, arr2)
newarr

array([1, 2, 3, 4, 5, 6])

Find intersection of the following two set arrays:



In [100]:
arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([3, 4, 5, 6])

newarr = np.intersect1d(arr1, arr2, assume_unique=True)
newarr

array([3, 4])

**Note**: the `intersect1d()` method takes an optional argument `assume_unique`, which if set to **True** can speed up computation. It should always be set to **True** when dealing with sets.



Find the difference of the set1 from set2:



In [101]:
set1 = np.array([1, 2, 3, 4])
set2 = np.array([3, 4, 5, 6])
newarr = np.setdiff1d(set1, set2, assume_unique=True)
newarr

array([1, 2])

**Note**: the `setdiff1d()` method takes an optional argument `assume_unique`, which if set to True can speed up computation. It should always be set to True when dealing with sets.



Find the symmetric difference of the set1 and set2:



In [102]:
set1 = np.array([1, 2, 3, 4])
set2 = np.array([3, 4, 5, 6])
newarr = np.setxor1d(set1, set2, assume_unique=True)
newarr

array([1, 2, 5, 6])

**Note**: the `setxor1d()` method takes an optional argument `assume_unique`, which if set to True can speed up computation. It should always be set to True when dealing with sets.

